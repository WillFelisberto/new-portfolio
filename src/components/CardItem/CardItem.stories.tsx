import { Meta, StoryObj } from '@storybook/react';
import CardItem from '.';

const meta: Meta<typeof CardItem> = {
  title: 'Components/CardItem',
  component: CardItem,
  parameters: {
    layout: 'fullscreen',
  },
};

export default meta;

type Story = StoryObj<typeof CardItem>;

export const EducationExample: Story = {
  args: {
    logoUrl: '/ufsc.jpg',
    logoAlt: 'Education Logo',
    title: 'Computer Science',
    subtitle: 'Bachelor of Science',
    dateRange: 'Jan 2015 • Dec 2019 • (4 years)',
    description: {
      lexicalData: {
        root: {
          direction: 'ltr',
          format: '',
          indent: 0,
          type: 'root',
          version: 1,
          children: [
            {
              detail: 0,
              format: 0,
              mode: 'normal',
              style: '',
              text: 'Description',
              type: 'text',
              version: 1,
            },
          ],
        },
      },
    },
    linkUrl: 'https://university.edu',
    linkText: 'University Name',
  },
};

export const ExperienceExample: Story = {
  args: {
    logoUrl: '/ufsc.jpg',
    logoAlt: 'Company Logo',
    title: 'Company Name',
    subtitle: 'Software Engineer',
    dateRange: 'Feb 2020 • Present • (3 years)',
    description: {
      lexicalData: {
        root: {
          direction: 'ltr',
          format: '',
          indent: 0,
          type: 'root',
          version: 1,
          children: [
            {
              detail: 0,
              format: 0,
              mode: 'normal',
              style: '',
              text: 'Description',
              type: 'text',
              version: 1,
            },
          ],
        },
      },
    },
    linkUrl: 'https://company.com',
    linkText: '@ Company Name',
    additionalContent: (
      <div>
        <p className="mb-3 mt-6 text-sm font-semibold text-gray-400">Skills</p>
        <div className="mb-8 flex flex-wrap gap-x-2 gap-y-3 lg:max-w-[350px]">
          <span className="rounded bg-gray-600 px-2 py-1 text-white">
            React
          </span>
          <span className="rounded bg-gray-600 px-2 py-1 text-white">
            Node.js
          </span>
        </div>
      </div>
    ),
  },
};
