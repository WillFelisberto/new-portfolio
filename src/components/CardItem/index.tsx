/* eslint-disable @typescript-eslint/no-explicit-any */
'use client';
import { fadeUpAnimation } from '@/lib/animations';
import { motion } from 'framer-motion';
import Image from 'next/image';
import RichTextComponent from '../RichText';

export type CardItemProps = {
  logoUrl: string;
  logoAlt: string;
  title: string;
  dateRange: string;
  subtitle?: string;
  description?: any; // RichText content
  linkUrl?: string;
  education?: boolean;
  additionalContent?: React.ReactNode;
};

const CardItem = ({
  logoUrl,
  logoAlt,
  title,
  subtitle,
  dateRange,
  description,
  linkUrl,
  education = false,
  additionalContent,
}: CardItemProps) => {
  return (
    <motion.div
      className="grid grid-cols-[40px,1fr] gap-4 md:gap-10"
      {...fadeUpAnimation}
      transition={{ duration: 0.5 }}
    >
      <div className="flex flex-col items-center gap-4">
        {education ? (
          <div className="h-[60px] w-[60px] rounded-full border border-gray-500 bg-gray-50 p-0.5">
            <Image
              src={logoUrl}
              width={60}
              height={60}
              className="h-[60px] w-[60px] rounded-full"
              alt={`${title}`}
            />
          </div>
        ) : (
          <div className="rounded-full border border-gray-500 p-0.5">
            <Image
              src={logoUrl}
              width={40}
              height={40}
              className="rounded-full"
              alt={logoAlt}
            />
          </div>
        )}

        <div className="h-full w-[1px] bg-gray-800" />
      </div>

      <div>
        <div className="flex flex-col gap-2 text-sm sm:text-base">
          {linkUrl ? (
            <a
              href={linkUrl}
              target="_blank"
              aria-label={title}
              className="text-sky-100 transition-colors hover:text-sky-500"
              rel="noreferrer"
            >
              {title}
            </a>
          ) : (
            <h4 className="text-gray-300">{title}</h4>
          )}
          {subtitle && <p className="text-gray-400">{subtitle}</p>}
          <span className="text-gray-500">{dateRange}</span>
          {description && <RichTextComponent lexicalData={description} />}
        </div>

        {additionalContent && <div className="mt-4">{additionalContent}</div>}
      </div>
    </motion.div>
  );
};

export default CardItem;
