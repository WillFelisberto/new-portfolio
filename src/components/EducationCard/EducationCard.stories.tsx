import { Meta, StoryObj } from '@storybook/react';
import EducationCard, { Education } from '.';

const meta: Meta<typeof EducationCard> = {
  title: 'Components/EducationCard',
  component: EducationCard,
  parameters: {
    layout: 'fullscreen',
  },
};

export default meta;

type Story = StoryObj<typeof EducationCard>;

const mockEducation: Education = {
  title: 'University of Technology',
  description:
    'Studied advanced computer science topics, including algorithms, data structures, and machine learning.',
  educationLogo: {
    url: '/ufsc.jpg',
  },
  url: 'https://university.tech',
  startDate: '2018-01-01',
  endDate: '2022-01-01',
  courseName: 'Computer Science',
};

export const Default: Story = {
  args: {
    education: mockEducation,
  },
};
