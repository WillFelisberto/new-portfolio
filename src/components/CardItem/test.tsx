import { render } from '@testing-library/react';

import CardItem from '.';

describe('<CardItem />', () => {
  const mockProps = {
    logoUrl: '/path/to/logo.png',
    logoAlt: 'Example Logo',
    title: 'Example Title',
    subtitle: 'Example Subtitle',
    dateRange: 'Jan 2020 • Dec 2023 • (3 years)',
    description: 'description',
    linkUrl: 'https://example.com',
    linkText: 'Example Link',
    additionalContent: <div>Additional Content</div>,
  };
  it('should render card with all required props correctly', () => {
    const { getByText, getByLabelText, getByAltText, container } = render(
      <CardItem {...mockProps} />,
    );

    expect(getByAltText(/Example Logo/i)).toBeInTheDocument();
    expect(getByText(/Example Title/i)).toBeInTheDocument();
    expect(getByLabelText(/Example Link/i)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
