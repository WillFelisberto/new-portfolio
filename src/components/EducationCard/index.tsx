import { formatDateRange } from '@/lib/utils';
import CardItem from '../CardItem';
import { DescriptionType } from '../RichText';

export type Education = {
  title: string;
  description: DescriptionType;
  educationLogo: {
    url: string;
  };
  url: string;
  startDate: string;
  endDate: string;
  courseName: string;
};

type EducationCardProps = {
  education: Education;
};

const EducationCard = ({ education }: EducationCardProps) => {
  const {
    startDate,
    endDate,
    description,
    courseName,
    educationLogo,
    title,
    url,
  } = education;

  const { formattedStartDate, formattedEndDate, formattedDuration } =
    formatDateRange(startDate, endDate);

  return (
    <CardItem
      logoUrl={educationLogo.url}
      logoAlt={`Logo da empresa ${title}`}
      subtitle={courseName}
      title={title}
      education
      dateRange={`${formattedStartDate} • ${formattedEndDate} • (${formattedDuration})`}
      description={description}
      linkUrl={url}
    />
  );
};

export default EducationCard;
