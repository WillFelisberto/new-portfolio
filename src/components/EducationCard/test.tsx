import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import EducationCard, { Education } from '.';
import { CardItemProps } from '../CardItem';

jest.mock('../CardItem', () => ({
  __esModule: true,
  default: ({ title, subtitle, dateRange, linkText }: CardItemProps) => (
    <div>
      <h1>{title}</h1>
      <h2>{subtitle}</h2>
      <p>{dateRange}</p>
      <a href="#">{linkText}</a>
    </div>
  ),
}));

jest.mock('@/lib/utils', () => ({
  formatDateRange: jest.fn(() => ({
    formattedStartDate: 'Jan 2020',
    formattedEndDate: 'Dec 2023',
    formattedDuration: '3 years',
  })),
}));

describe('<EducationCard />', () => {
  const mockEducation: Education = {
    title: 'Bachelor of Science',
    description: {
      lexicalData: {
        root: {
          direction: 'ltr',
          format: '',
          indent: 0,
          type: 'root',
          version: 1,
          children: [
            {
              detail: 0,
              format: 0,
              mode: 'normal',
              style: '',
              text: 'Studied advanced computer science topics, including algorithms, data structures, and machine learning.',
              type: 'text',
              version: 1,
            },
          ],
        },
      },
    },
    educationLogo: {
      url: '/path/to/logo.png',
    },
    url: 'https://university.edu',
    startDate: '2020-01-01',
    endDate: '2023-12-31',
    courseName: 'Information Technology',
  };

  it('renders the education details correctly', () => {
    const { container } = render(<EducationCard education={mockEducation} />);

    // Verifica se os dados do curso e instituição estão sendo renderizados
    expect(screen.getByText(/Information Technology/i)).toBeInTheDocument();
    expect(screen.getByText(/Bachelor of Science/i)).toBeInTheDocument();

    // Verifica o intervalo de datas formatado
    expect(
      screen.getByText('Jan 2020 • Dec 2023 • (3 years)'),
    ).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
