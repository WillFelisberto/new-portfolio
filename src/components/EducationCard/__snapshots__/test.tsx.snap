// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EducationCard /> renders the education details correctly 1`] = `
<div>
  <h1>
    Information Technology
  </h1>
  <h2 />
  <p>
    Jan 2020 • Dec 2023 • (3 years)
  </p>
  <a
    href="#"
  >
    Go to Bachelor of Science
  </a>
</div>
`;
